### PR TITLE
feat: add login endpoint

### PR DESCRIPTION
**Pull request by shawnhankim**
**Issue:**
- https://github.com/nginxinc/nginx-openid-connect/issues/73

**Background:**
As a Product Manager,

I want a SPA to start a landing page of Single Page App(SPA) without OpenID Connect (OIDC) flow. In the landing page, I want a SPA to just show default data by calling public APIs without API authorization. 

Afterwards, I want to start the OIDC flow when I just click `login` button to display information in detail by calling private APIs based on API authorization using an access token.

**Description:**
- Added a `/login` location to start OIDC flow.
- Added a map variable of `$oidc_landing_page` to determine where to send browser after successful login from the IdP.

### CODE REVIEW

**Comment by shawnhankim** [comment]
@route443 :
- Thanks for your review in detail for the [PR](https://github.com/nginxinc/nginx-openid-connect/pull/62).
- This PR is to simplify from the previous [PR](https://github.com/nginxinc/nginx-openid-connect/pull/68).
- For you to easily manage this repo to reduce any concerns from the enhancements based on the reviews on the [PR](https://github.com/nginxinc/nginx-openid-connect/pull/62), I have divided a big PR into small PRs, and this is one of PRs.
- Let me know if you have any questions.

**Review by route443 — CHANGES_REQUESTED** [review]
Hi Shawn. Thank you for your contribution to our nginx oidc project.
As I said in previous PRs, we are not using Conventional Commits for this project. So I wouldn't use this approach unnecessarily. Please change the commit name.

File: frontend.conf [hunk 1]
```diff
@@ -31,6 +31,20 @@ server {
         
         access_log /var/log/nginx/access.log main_jwt;
     }
+
+    location = /login {
```
[the remark this answers is not in the document]
Comment (shawnhankim) [reply frontend.conf]: This endpoint requirement was originally requested by PMs of NGINX Management Suite - API Connectivity Manager (a.k.a. NMS-ACM) as follows. In the meantime, I have added another common purpose.

**Product Requirements of NMS-ACM:**
- **Problem**: Current NJS implementation doesn’t have login endpoints for client apps (SPA) to interact with. Client Apps require /login function as part of relying party when user clicks on login button. 
- **Change Required**: Expose the `/login` nginx location block here (`openid_connect.server_conf`) & proxy it to existing authz endpoint configured in the map variable(`openid_connect_configuration.conf`) of  `$oidc_authz_endpoint`. This would outsource the login function to IdP as its configured. 

**Reason why we need regardless for the general purpose of NMS-ACM requirement:**
- The current example only shows the scenario where `/` location always need to start with user authentication.
- There are many web sites (which are proxied from `/` location) that do not have to be started with user authentication by using IdP's login GUI.
- For those cases, this endpoint can be used by the SPA when users click on login button.

**To Sum Up:**
- To prevent confusion with an additional example, I have removed this in this `frontend.conf`, and the endpoint has been moved into `openid_connect_configuration.conf` that can be used by customers based on their requirements.
- Let me know if this reason is not enough for you.

File: openid_connect.js [hunk 2]
```diff
@@ -253,6 +253,15 @@ function validateIdToken(r) {
     }
 }
 
+// Redirect URI after successful login from the OP.
+function redirectPostLogin(r) {
```
[the remark this answers is not in the document]
Comment (shawnhankim) [reply openid_connect.js:257]: Oh, yeah! I have moved this into the function of code exchange. Let me know if I miss anything.

File: openid_connect_configuration.conf [hunk 3]
```diff
@@ -44,6 +44,15 @@ map $host $oidc_scopes {
     default "openid+profile+email+offline_access";
 }
 
+map $host $oidc_landing_page {
+    # Where to send browser after successful login. This option is only
```
[the remark this answers is not in the document]
Comment (shawnhankim) [reply openid_connect_configuration.conf]: Thanks! Addressed it.

File: frontend.conf (hunk 1 above)
[the remark this answers is not in the document]
Comment (route443) [reply frontend.conf]: The list of reasons is not very convincing. 
1. JWT authentication in NGINX works in such a way that if you specify `/login` location, then only it will be protected. The principle of "explicit allow" is a security principle that states that only actions or activities that have been explicitly permitted should be allowed to take place. This means that anything that is not specifically allowed is prohibited.
2. The `openid_connect.server_conf` file contains OIDC functional locations, that is, by default, the user does not need to change anything in it.
3. We work as a proxy, that is, the more locations we have, the less predictable the end application can be, simply because locations can overlap.

Your `/login` example is app specific, that is, nothing prevents you from creating the configuration that NMS-ACM need, we, in turn, pursue the goal of being as predictable (simple) as possible.
We don't make a Swiss Army knife for all occasions, we provide a simple solution that should meet the standard OIDC requirements and not be overly complicated.

[the remark this answers is not in the document]
Comment (shawnhankim) [reply frontend.conf]: - FYI. **This feature was originally requested by the PM** in the team who **managed this OIDC repo** before you started working on this repo. Hence, it was firstly implemented into the NGINX Management Suite - API Connectivity Manager, and the PM's plan was to donate it here to be used by more customers.
- However, I totally understand what you point out, and we may **need to find another repo** to add **more app specific solutions**, because the purpose of this repo is just to provide a simple reference implementation. So we can provide much more convenient and easy solutions to customers to influence this OIDC solution over the world.
- Let me think of how to provide the simpler solution of the feature here. I appreciate your point.

[the remark this answers is not in the document]
Comment (shawnhankim) [reply frontend.conf]: @route443 :
* The `/login` endpoint has been removed as this repo is only to provide reference implementation for a simple proxy.
* The `$oidc_landing_page` is only kept as I like your suggestion of variable name.
* The application level reference implementation will be separately considered in other repo as this repo is only for simple implementation for a proxy after discussing with the PM of NGINX Management Suite.
* Feel free to let me know if you have other thoughts.
